### ISSUE TITLE
Document loading pcmemtest from UEFI grub2

### ISSUE THREAD

**Issue by gordonmessmer** [martinwhitaker/pcmemtest#9]
The documentation currently states "Either image can be booted by an intermediate bootloader using the Linux 32-bit, 64-bit, or 64-bit EFI handover boot protocols," but I have been unable to successfully load pcmemtest.efi from GRUB2.  GRUB2 will report that it is loading the image, and then appears to halt with that message on screen.  Perhaps I'm not setting the required graphics settings?  Would you provide an example GRUB2 stanza for UEFI to the documentation?

**Comment by martinwhitaker** [comment]
I don't use GRUB2 on my systems. But try
````
set gfxpayload 640x400,800x600,1024x768
linux /boot/memtest.efi
````
(assuming `/boot/memtest.efi` is the correct path on your system)

**Comment by gordonmessmer** [comment]
I've tried various combinations of

    set gfxpayload=keep
    load_video

and

    set     gfxpayload=640x400,800x600,1024x768

with "linux" and "linuxefi /\<path\>/pcmemtest.efi".  All of these have the same result: GRUB appears to load the image (which is to say that it doesn't print an error indicating that the image can't be found) but the system seems to halt.  

So far the only way I've been able to run pcmemtest is to add the image directly to the UEFI boot list using efibootmgr, so I'm pretty sure the image itself works.  I just can't seem to load it from GRUB2.

**Comment by martinwhitaker** [comment]
Try changing the line
````
#define DEBUG 0
````
near the top of the boot/efisetup.c file to
````
#define DEBUG 1
````
and rebuilding. Does that print some debug lines when you use the `linuxefi` command to start pcmemtest?

As you mention the `linuxefi` command, I assume you are using Fedora, who heavily modify GRUB2.

**Comment by gordonmessmer** [comment]
Yes, this is Fedora.

With DEBUG 1, I see a prompt with information about the graphics when I load pcmemtest directly from UEFI, and nothing when I try to load it from GRUB2.

**Comment by gordonmessmer** [comment]
I've tested GRUB2 on Debian to eliminate problems caused by Fedora's modifications.  On Debian, if I use "linux /boot/pcmemtest.efi" or "linuxefi /boot/pcmemtest.efi", the system will either reset or using 100% of a CPU core with no visible output.  So, I don't seem to be able to use the EFI handover boot protocol from either Debian or Fedora GRUB.

On the other hand, both Debian and (I'm told) openSUSE GRUB can load pcmemtest.efi using "chainloader /boot/pcmemtest.efi", so Fedora's GRUB modifications might be interfering with that option.

I think we can close this issue for now, but I'd like to ask if you'd accept new issues for a) testing the EFI handover boot protocol under GRUB or b) adding certificate table headers so that the EFI binary can be signed.

**Comment by martinwhitaker** [comment]
I know the EFI handover boot protocol works with GRUB, because I've tested it. You can confirm that by downloading and testing one of the Mageia ISOs from https://www.mageia.org/en/downloads/. By default they use the GRUB "linux32" command to boot using the 32-bit boot protocol (because that works with both EFI and legacy boot), but you can edit the boot menu entry and change "linux32" to "linuxefi" to test the EFI handover protocol.

I've just added a grub-iso target in build64/Makefile, so you can try building a test ISO yourself. This provides boot menu options to use any of the "linux", "linux16", "linux32", or "linuxefi" commands. See the comment in the make file for more details. All options work when built in Mageia 8. You may need to override the GRUB_FONT_DIR, GRUB_LIB_DIR, and GRUB_MKIMAGE variables when building in a different distribution.

I have no personal interest in signed binaries, but am happy to accept patches.

**Comment by raddirad** [comment]
here's how we start pcmemtest via UEFI netboot with grub

This is just the named pipe that gets transfered to the client

```
set timeout=0
menuentry 'Starting pcmemtest' {
set gfxpayload=keep
linux (pxe)/linux//memtest.efi
}
```

**Comment by gordonmessmer** [comment]
Thanks.  I do see that the Mageia boot ISO works with "linux32" and "linux".  I'm not sure why "linux" doesn't work on Debian 11, and "linux32" isn't available on my Debian 11 system, but "chainload" does work, and I can start testing from there.

I think af8e071 more than satisfies this issue, so let's close this.

**Comment by martinwhitaker** [comment]
I just installed Debian 11 in a VM and used it to build grub-memtest.iso. I tested both booting from an emulated CD (in VirtualBox) and booting from a USB stick (on real hardware). Both "linux" and "linux16" worked when booted in legacy mode, and both "linux" and "linuxefi" worked when booted in EFI mode. "linux32" is not available with Debian GRUB - that's a command I added in the Mageia GRUB package after Mageia adopted the Fedora patches that make "linux" an alias to either "linux16" or "linuxefi".

The only thing I can think of is that GRUB is passing on invalid video mode information on your particular hardware. I believe the "linuxefi" command ignores the "gfxpayload" setting, and passes on the size of the EFI frame buffer (that's the behaviour I see with grub-memtest.iso). Have you tested it on more than one machine?
